File: web/src/types/alert.ts
import { v4 as uuidv4 } from 'uuid';

export enum AlertType {
  DrowsyDriver = 'DrowsyDriver',
  LaneDeparture = 'LaneDeparture',
  CargoTamper = 'CargoTamper',
  LicensePlateMatch = 'LicensePlateMatch',
  WeatherHazard = 'WeatherHazard',
  HighTemperature = 'HighTemperature',
  LowDiskSpace = 'LowDiskSpace',
  HighCpuUsage = 'HighCpuUsage',
  NetworkFailure = 'NetworkFailure',
  SensorFailure = 'SensorFailure',
  HarshBraking = 'HarshBraking',
  RapidAcceleration = 'RapidAcceleration',
  SeatbeltNotFastened = 'SeatbeltNotFastened',
  DoorOpenWhileMoving = 'DoorOpenWhileMoving',
  OverSpeeding = 'OverSpeeding',
  UpdateAvailable = 'UpdateAvailable',
  UpdateFailed = 'UpdateFailed',
  RollbackTriggered = 'RollbackTriggered',
  ConfigError = 'ConfigError',
}

export enum AlertSeverity {
  Info = 'Info',
  Warning = 'Warning',
  Critical = 'Critical',
  Emergency = 'Emergency',
}

export enum ActionType {
  TriggerBuzzer = 'TriggerBuzzer',
  FlashLed = 'FlashLed',
  ShowOnDisplay = 'ShowOnDisplay',
  SendCanMessage = 'SendCanMessage',
  ActivateRelay = 'ActivateRelay',
  LogToServer = 'LogToServer',
  SendSms = 'SendSms',
}

export enum AlertStatus {
  Triggered = 'Triggered',
  Acknowledged = 'Acknowledged',
  Resolved = 'Resolved',
  Suppressed = 'Suppressed',
}

export interface Alert {
  id: string;
  alert_id: string;
  truck_id: string;
  alert_type: AlertType;
  severity: AlertSeverity;
  message: string;
  triggered_at: string;
  acknowledged_at: string | null;
  resolved_at: string | null;
  source: string;
  context: any; // JSON object with alert-specific context
  actions: AlertAction[];
  status: AlertStatus;
  created_at: string;
  updated_at: string;
}

export interface AlertAction {
  action_id: string;
  action_type: ActionType;
  target: string;
  parameters: any; // JSON object with action-specific parameters
  executed_at: string | null;
  success: boolean;
  error: string | null;
}

export interface AlertSummary {
  id: string;
  alert_id: string;
  truck_id: string;
  alert_type: AlertType;
  severity: AlertSeverity;
  message: string;
  triggered_at: string;
  status: AlertStatus;
  truck_license_plate: string;
  truck_model: string;
  truck_make: string;
}

export interface AlertStats {
  total_alerts: number;
  active_alerts: number;
  acknowledged_alerts: number;
  resolved_alerts: number;
  by_severity: Record<string, number>;
  by_type: Record<string, number>;
  by_truck: Record<string, number>;
  last_24_hours: number;
  last_7_days: number;
  last_30_days: number;
}

export interface DashboardAlerts {
  recent_alerts: AlertSummary[];
  stats: AlertStats;
  top_trucks: [string, number][];
  alerts_by_type: Record<string, number>;
  alerts_by_severity: Record<string, number>;
  time_range_hours: number;
}

File: truck-central-server/src/models/alert.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use chrono::{DateTime, Utc};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Alert {
    pub id: Uuid,
    pub alert_id: String,
    pub truck_id: Uuid,
    pub alert_type: AlertType,
    pub severity: AlertSeverity,
    pub message: String,
    pub triggered_at: DateTime<Utc>,
    pub acknowledged_at: Option<DateTime<Utc>>,
    pub resolved_at: Option<DateTime<Utc>>,
    pub source: String,
    pub context: serde_json::Value,
    pub actions: Vec<AlertAction>,
    pub status: AlertStatus,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum AlertType {
    DrowsyDriver,
    LaneDeparture,
    CargoTamper,
    LicensePlateMatch,
    WeatherHazard,
    HighTemperature,
    LowDiskSpace,
    HighCpuUsage,
    NetworkFailure,
    SensorFailure,
    HarshBraking,
    RapidAcceleration,
    SeatbeltNotFastened,
    DoorOpenWhileMoving,
    OverSpeeding,
    UpdateAvailable,
    UpdateFailed,
    RollbackTriggered,
    ConfigError,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum AlertSeverity {
    Info,
    Warning,
    Critical,
    Emergency,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlertAction {
    pub action_id: String,
    pub action_type: ActionType,
    pub target: String,
    pub parameters: serde_json::Value,
    pub executed_at: Option<DateTime<Utc>>,
    pub success: bool,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum ActionType {
    TriggerBuzzer,
    FlashLed,
    ShowOnDisplay,
    SendCanMessage,
    ActivateRelay,
    LogToServer,
    SendSms,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum AlertStatus {
    Triggered,
    Acknowledged,
    Resolved,
    Suppressed,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlertSummary {
    pub id: Uuid,
    pub alert_id: String,
    pub truck_id: Uuid,
    pub alert_type: AlertType,
    pub severity: AlertSeverity,
    pub message: String,
    pub triggered_at: DateTime<Utc>,
    pub status: AlertStatus,
    pub truck_license_plate: String,
    pub truck_model: String,
    pub truck_make: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlertStats {
    pub total_alerts: i64,
    pub active_alerts: i64,
    pub acknowledged_alerts: i64,
    pub resolved_alerts: i64,
    pub by_severity: std::collections::HashMap<String, i64>,
    pub by_type: std::collections::HashMap<String, i64>,
    pub by_truck: std::collections::HashMap<Uuid, i64>,
    pub last_24_hours: i64,
    pub last_7_days: i64,
    pub last_30_days: i64,
}